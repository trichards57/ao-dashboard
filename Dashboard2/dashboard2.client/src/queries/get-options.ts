import { queryOptions } from "@tanstack/react-query";
import { notFound } from "@tanstack/react-router";

export default function getOptions<T>(uri: string, queryKey: string[]) {
  return queryOptions({
    queryKey,
    queryFn: async () => {
      const response = await fetch(uri, {
        method: "GET",
        headers: {
          "Content-Type": "application/json",
        },
      });

      if (!response.ok) {
        if (response.status === 404) {
          throw notFound();
        }
        throw new Error("Network error");
      }

      return response.json() as Promise<T>;
    },
    staleTime: 10 * 60 * 1000,
    throwOnError: true,
  });
}
