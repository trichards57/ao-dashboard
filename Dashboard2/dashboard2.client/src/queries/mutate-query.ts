import { useMutation, useQueryClient } from "@tanstack/react-query";
import { notFound } from "@tanstack/react-router";

export default function useUpdate(uri: string, queryKey: string[]) {
  const queryClient = useQueryClient();

  return function useUpdateInner<TBody>(id: string) {
    return useMutation({
      mutationFn: async (body: TBody) => {
        const response = await fetch(`${uri}/${id}`, {
          method: "POST",
          headers: {
            "Content-Type": "application/json",
          },
          body: JSON.stringify(body),
        });

        if (!response.ok) {
          if (response.status === 404) {
            // eslint-disable-next-line @typescript-eslint/no-throw-literal
            throw notFound();
          }
          throw new Error("Failed to update.");
        }
      },
      onSuccess: () => {
        queryClient.invalidateQueries({ queryKey });
      },
    });
  };
}
