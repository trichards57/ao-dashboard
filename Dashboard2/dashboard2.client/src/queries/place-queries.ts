import {
  QueryClient,
  queryOptions,
  useSuspenseQuery,
} from "@tanstack/react-query";
import { notFound } from "@tanstack/react-router";

export type Region =
  | "All"
  | "EastOfEngland"
  | "EastMidlands"
  | "London"
  | "NorthEast"
  | "NorthWest"
  | "SouthEast"
  | "SouthWest"
  | "WestMidlands";

export function districtOptions(region: Region) {
  return queryOptions({
    queryKey: ["districts", region],
    queryFn: async () => {
      if (region == "All") {
        return [];
      }

      const response = await fetch(`/api/places/${region}`, {
        method: "GET",
        headers: {
          "Content-Type": "application/json",
        },
      });

      if (!response.ok) {
        if (response.status === 404) {
          throw notFound();
        }
        throw new Error("Failed to fetch districts.");
      }

      return response.json() as Promise<string[]>;
    },
    staleTime: 10 * 60 * 1000,
  });
}

export function hubOptions(region: Region, district: string) {
  return queryOptions({
    queryKey: ["hubs", region, district],
    queryFn: async () => {
      if (region == "All" || district.toUpperCase() == "ALL") {
        return [];
      }

      const response = await fetch(`/api/places/${region}/${district}`, {
        method: "GET",
        headers: {
          "Content-Type": "application/json",
        },
      });

      if (!response.ok) {
        if (response.status === 404) {
          throw notFound();
        }
        throw new Error("Failed to fetch hubs.");
      }

      return response.json() as Promise<string[]>;
    },
    staleTime: 10 * 60 * 1000,
  });
}

export function useDistricts(region: Region) {
  return useSuspenseQuery(districtOptions(region));
}

export function useHubs(region: Region, district: string) {
  return useSuspenseQuery(hubOptions(region, district));
}

export function preloadDistricts(queryClient: QueryClient, region: Region) {
  return queryClient.ensureQueryData(districtOptions(region));
}

export function preloadHubs(
  queryClient: QueryClient,
  region: Region,
  district: string,
) {
  return queryClient.ensureQueryData(hubOptions(region, district));
}
