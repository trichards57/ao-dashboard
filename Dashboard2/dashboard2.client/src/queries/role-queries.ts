import { QueryClient, useSuspenseQuery } from "@tanstack/react-query";

import getOptions from "./get-options";
import useUpdate from "./mutate-query";

export type Permission = "Deny" | "Read" | "Write";

export interface RolePermissions {
  id: string;
  name: string;
  permissions: Permission;
  vehicleConfiguration: Permission;
  vorData: Permission;
}

export interface RolePermissionsUpdate {
  permissions: Permission;
  vehicleConfiguration: Permission;
  vorData: Permission;
}

export function useUpdateRole(id: string) {
  return useUpdate("/api/roles", ["role"])<RolePermissionsUpdate>(id);
}

export function roleOptions(id: string) {
  return getOptions<RolePermissions>(`/api/roles/${id}`, ["role", id]);
}

export const allRoleOptions = getOptions<RolePermissions[]>("/api/roles", [
  "role",
]);

export function useRole(id: string) {
  return useSuspenseQuery(roleOptions(id));
}

export function useRoles() {
  return useSuspenseQuery(allRoleOptions);
}

export function preloadRole(queryClient: QueryClient, id: string) {
  return queryClient.ensureQueryData(roleOptions(id));
}

export function preloadRoles(queryClient: QueryClient) {
  return queryClient.ensureQueryData(allRoleOptions);
}
