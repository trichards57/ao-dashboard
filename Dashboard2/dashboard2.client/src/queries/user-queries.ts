import {
  useMutation,
  useQueryClient,
  UseSuspenseQueryOptions,
} from "@tanstack/react-query";
import { notFound } from "@tanstack/react-router";

export interface UserInfo {
  realName: string;
  userId: string;
  email: string;
  role: string;
  amrUsed: string;
  lastAuthenticated: string;
  otherClaims: Record<string, string>;
}

export interface UserWithRole {
  id: string;
  name: string;
  roleId: string;
  role: string;
}

export interface UserRoleUpdate {
  roleId: string;
}

export const useUpdateUser = (id: string) => {
  const queryClient = useQueryClient();

  return useMutation({
    mutationFn: async (user: UserRoleUpdate) => {
      const response = await fetch(`/api/users/${id}`, {
        method: "POST",
        headers: {
          "Content-Type": "application/json",
        },
        body: JSON.stringify(user),
      });

      if (!response.ok) {
        if (response.status === 404) {
          throw notFound();
        }
        throw new Error("Failed to update user.");
      }
    },
    onSuccess: () => {
      queryClient.invalidateQueries({ queryKey: ["user"] });
    },
  });
};

export const userSettings: (
  id: string,
) => UseSuspenseQueryOptions<UserWithRole> = (id: string) => ({
  queryKey: ["user", id],
  queryFn: async () => {
    const response = await fetch(`/api/users/${id}`, {
      method: "GET",
      headers: {
        "Content-Type": "application/json",
      },
    });

    if (!response.ok) {
      if (response.status === 404) {
        throw notFound();
      }
      throw new Error("Failed to fetch user settings.");
    }

    return response.json() as Promise<UserWithRole>;
  },
  staleTime: 10 * 60 * 1000
});

export const allUserOptions: UseSuspenseQueryOptions<UserWithRole[]> = {
  queryKey: ["user"],
  queryFn: async () => {
    const response = await fetch("/api/users", {
      method: "GET",
      headers: {
        "Content-Type": "application/json",
      },
    });

    if (!response.ok) {
      throw new Error("Failed to fetch users.");
    }

    return response.json() as Promise<UserWithRole[]>;
  },
  staleTime: 10 * 60 * 1000
};

export const userMeOptions: UseSuspenseQueryOptions<UserInfo | null> = {
  queryKey: ["me"],
  queryFn: async () => {
    const response = await fetch("/api/users/me", {
      method: "GET",
      headers: {
        "Content-Type": "application/json",
      },
    });

    if (!response.ok) {
      return Promise.resolve(null);
    }

    return response.json() as Promise<UserInfo>;
  },
  staleTime: 1000 * 60 * 5,
};
