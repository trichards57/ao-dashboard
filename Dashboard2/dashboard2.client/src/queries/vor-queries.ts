import { QueryClient, useSuspenseQuery } from "@tanstack/react-query";
import getOptions from "./get-options";

export interface VorStatistics {
  totalVehicles: number;
  availableVehicles: number;
  vorVehicles: number;
  pastAvailability: Record<string, number>;
}

export interface VorStatus {
  id: string;
  region: string;
  hub: string;
  district: string;
  registration: string;
  callSign: string;
  summary: string;
  isVor: boolean;
  dueBack: string;
}

export function statisticsOptions(
  region: string,
  district: string,
  hub: string,
) {
  return getOptions<VorStatistics>(
    `/api/vor/statistics?region=${region}&district=${district}&hub=${hub}`,
    ["statistics", region, district, hub],
  );
}

export function useStatistics(region: string, district: string, hub: string) {
  return useSuspenseQuery(statisticsOptions(region, district, hub));
}

export function preloadStatistics(
  queryClient: QueryClient,
  region: string,
  district: string,
  hub: string,
) {
  return queryClient.ensureQueryData(statisticsOptions(region, district, hub));
}

export function statusOptions(
  region: string | undefined,
  district: string | undefined,
  hub: string | undefined,
) {
  return getOptions<VorStatus[]>(
    `/api/vor?region=${region ?? "All"}&district=${district ?? "All"}&hub=${hub ?? "All"}`,
    ["status", region ?? "All", district ?? "All", hub ?? "All"],
  );
}

export function useStatus(
  region: string | undefined,
  district: string | undefined,
  hub: string | undefined,
) {
  return useSuspenseQuery(statusOptions(region, district, hub));
}

export function preloadStatus(
  queryClient: QueryClient,
  region: string | undefined,
  district: string | undefined,
  hub: string | undefined,
) {
  return queryClient.ensureQueryData(statusOptions(region, district, hub));
}
