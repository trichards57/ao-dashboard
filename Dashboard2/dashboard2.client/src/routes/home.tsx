import { createFileRoute } from '@tanstack/react-router'

export const Route = createFileRoute('/home')({
  component: () => <div>Hello /home!</div>
})