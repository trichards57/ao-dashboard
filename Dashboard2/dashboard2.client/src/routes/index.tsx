import { createFileRoute } from "@tanstack/react-router";

import useTitle from "../components/useTitle";
import { redirectIfLoggedIn } from "../support/check-logged-in";

export function Index() {
  useTitle();

  return (
    <section className="hero">
      <div className="hero-body">
        <p className="title">Welcome</p>
        <p className="subtitle">
          Welcome to the AO Ambulance Dashboard. You will need to log in to use
          this application.
        </p>
      </div>
      <div className="hero-foot is-flex is-justify-content-center">
        <a className="button is-primary" href="/Identity/Account/Login">
          Log In
        </a>
      </div>
    </section>
  );
}

export const Route = createFileRoute("/")({
  beforeLoad: ({ context }) => redirectIfLoggedIn(context),
  component: Index,
});
