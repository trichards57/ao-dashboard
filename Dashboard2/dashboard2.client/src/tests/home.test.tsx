import { render } from "@testing-library/react";
import { act } from "react";
import ReactDOM from "react-dom/client";
import { Mock, beforeEach, describe, expect, it, vi } from "vitest";

import { useDistricts, useHubs } from "../queries/place-queries";
import { VorStatistics, useStatistics } from "../queries/vor-queries";
import { Home } from "../routes/home";

vi.mock("../queries/vor-queries");
vi.mock("../queries/place-queries");

const testItem: VorStatistics = {
  availableVehicles: 10,
  vorVehicles: 5,
  totalVehicles: 15,
  pastAvailability: {
    "2020-01-01": 10,
    "2020-02-01": 11,
    "2020-03-01": 12,
    "2020-04-01": 13,
    "2020-05-01": 14,
    "2020-06-01": 15,
    "2020-07-01": 14,
    "2020-08-01": 13,
    "2020-09-01": 12,
    "2020-10-01": 11,
    "2020-11-01": 10,
    "2020-12-01": 9,
  },
};

const testDistricts = ["Avon", "Bedfordshire", "Berkshire", "Buckinghamshire"];
const testHubs = ["Bristol", "Bath", "Weston-super-Mare", "Yeovil"];

describe("Home Component", () => {
  beforeEach(() => {
    (useStatistics as Mock).mockReturnValue({
      data: testItem,
    });
    (useDistricts as Mock).mockReturnValue({
      data: testDistricts,
    });
    (useHubs as Mock).mockReturnValue({
      data: testHubs,
    });
  });

  it("renders without error", () => {
    const root = ReactDOM.createRoot(document.createElement("div"));
    expect(() => {
      act(() => {
        root.render(<Home region="All" district="All" hub="All" />);
        root.unmount();
      });
    }).not.toThrow();
  });

  it("renders correctly", () => {
    const { asFragment } = render(
      <Home region="SouthWest" district="Avon" hub="Bristol" />,
    );
    expect(asFragment()).toMatchSnapshot();
    expect(useStatistics).toHaveBeenCalledWith("SouthWest", "Avon", "Bristol");
    expect(useDistricts).toHaveBeenCalledWith("SouthWest");
    expect(useHubs).toHaveBeenCalledWith("SouthWest", "Avon");
  });
});
