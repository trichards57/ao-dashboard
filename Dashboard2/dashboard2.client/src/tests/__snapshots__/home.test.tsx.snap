// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Home Component > renders correctly 1`] = `
<DocumentFragment>
  <div
    class="columns is-desktop"
  >
    <div
      class="column"
    >
      <div
        class="field"
      >
        <label
          class="label"
          for="region"
        >
          Region
        </label>
        <div
          class="select is-fullwidth"
        >
          <select
            id="region"
          >
            <option
              value="All"
            >
              All
            </option>
            <option
              value="EastOfEngland"
            >
              East of England
            </option>
            <option
              value="EastMidlands"
            >
              East Midlands
            </option>
            <option
              value="London"
            >
              London
            </option>
            <option
              value="NorthEast"
            >
              North East
            </option>
            <option
              value="NorthWest"
            >
              North West
            </option>
            <option
              value="SouthEast"
            >
              South East
            </option>
            <option
              value="SouthWest"
            >
              South West
            </option>
            <option
              value="WestMidlands"
            >
              West Midlands
            </option>
          </select>
        </div>
      </div>
    </div>
    <div
      class="column"
    >
      <div
        class="field"
      >
        <label
          class="label"
          for="district"
        >
          District
        </label>
        <div
          class="select is-fullwidth"
        >
          <select
            id="region"
          >
            <option
              value="All"
            >
              All
            </option>
            <option
              value="Avon"
            >
              Avon
            </option>
            <option
              value="Bedfordshire"
            >
              Bedfordshire
            </option>
            <option
              value="Berkshire"
            >
              Berkshire
            </option>
            <option
              value="Buckinghamshire"
            >
              Buckinghamshire
            </option>
          </select>
        </div>
      </div>
    </div>
    <div
      class="column"
    >
      <div
        class="field"
      >
        <label
          class="label"
          for="hub"
        >
          Hub
        </label>
        <div
          class="select is-fullwidth"
        >
          <select
            id="hub"
          >
            <option
              value="all"
            >
              All
            </option>
            <option
              value="Bristol"
            >
              Bristol
            </option>
            <option
              value="Bath"
            >
              Bath
            </option>
            <option
              value="Weston-super-Mare"
            >
              Weston-super-Mare
            </option>
            <option
              value="Yeovil"
            >
              Yeovil
            </option>
          </select>
        </div>
      </div>
    </div>
  </div>
  <div
    class="columns"
  >
    <div
      class="column"
    >
      <h2
        class="subtitle is-5"
      >
        Vehicle Availability
      </h2>
      <div
        class="vehicle-availability-container"
      >
        <canvas
          height="0"
          id="availability-chart"
          role="img"
          style="display: block; box-sizing: border-box; height: 0px; width: 0px;"
          width="0"
        />
      </div>
    </div>
    <div
      class="column"
    >
      <h2
        class="subtitle is-5"
      >
        Historic Availability
      </h2>
      <div
        class="historic-availability-container"
      >
        <canvas
          height="0"
          id="historic-data-chart"
          role="img"
          style="display: block; box-sizing: border-box; height: 0px; width: 0px;"
          width="0"
        />
      </div>
    </div>
  </div>
</DocumentFragment>
`;
